### PR TITLE
Dense constraints in the maplab server

### PR DESCRIPTION
**Pull request by LBern**
## General
This PR contains a few additions to the server. First, a tiny improvement to the dense parallel process such that it won't _crash'n'burn_ any longer if the number of elements is less than the number available threads. Second, it adds the dense mapping constraint computation to the maplab server. 

**Example mission:**
Screenshot from 2020-07-30 16-54-14

**Vertices are now connected by LC edges coming from the dense constraints:**
Screenshot from 2020-07-30 16-55-16

## Changelog
 * Improved dense parallel process
 * Added gtests for the dense parallel process
 * Added dense mapping constraints to the maplab server
 * Adapted arche test launch files to include a point cloud map

### CODE REVIEW

File: applications/maplab-server-node/share/arche/arche-maplab-server-minimal.rviz [hunk 1]
```diff
@@ -266,17 +200,90 @@ Visualization Manager:
       Marker Topic: /artifact_manager_node/artifact_visualization
       Name: Artifact Observations
       Namespaces:
-        {}
+        arche_1 - A1 - baby goat: true
+        arche_1 - A2 - goat: true
+        arche_1 - A6 - green sheep: true
+        arche_2 - A3 - red sheep: true
+        arche_2 - A4 - blue sheep: true
+        arche_2 - A5 - black sheep: true
```
Comment (floriantschopp) [inline applications/maplab-server-node/share/arche/arche-maplab-server-minimal.rviz:208]: wtf is this :-D

File: applications/maplab-server-node/share/euroc/rviz/euroc-maplab-server-minimal.rviz [hunk 2]
```diff
@@ -30,10 +30,6 @@ Panels:
     Name: Time
     SyncMode: 0
     SyncSource: Image
-  - Class: mav_planning_rviz/PlanningPanel
```
Comment (floriantschopp) [inline applications/maplab-server-node/share/euroc/rviz/euroc-maplab-server-minimal.rviz:33]: +1

File: applications/maplab-server-node/src/maplab-server-node.cc [hunk 3]
```diff
@@ -207,29 +211,38 @@ void MaplabServerNode::start() {
...
 
-  LOG(INFO) << "[MaplabServerNode] Stopping MapMerging thread...";
-  if (submap_merging_thread_.joinable()) {
-    submap_merging_thread_.join();
+  try {
+    LOG(INFO) << "[MaplabServerNode] Stopping MapMerging thread...";
+    if (submap_merging_thread_.joinable()) {
+      submap_merging_thread_.join();
+    }
+    LOG(INFO) << "[MaplabServerNode] Done.";
```
Comment (floriantschopp) [inline applications/maplab-server-node/src/maplab-server-node.cc:223]:
```suggestion
    LOG(INFO) << "[MaplabServerNode] Done stopping MapMerging thread.";
```

File: applications/maplab-server-node/src/maplab-server-node.cc [hunk 4]
```diff
@@ -207,29 +211,38 @@ void MaplabServerNode::start() {
...
 
-  LOG(INFO) << "[MaplabServerNode] Stopping SubmapProcessing threads...";
-  submap_loading_thread_pool_.stop();
-  submap_loading_thread_pool_.waitForEmptyQueue();
-  LOG(INFO) << "[MaplabServerNode] Done.";
+  try {
+    LOG(INFO) << "[MaplabServerNode] Stopping SubmapProcessing threads...";
+    submap_loading_thread_pool_.stop();
+    submap_loading_thread_pool_.waitForEmptyQueue();
+    LOG(INFO) << "[MaplabServerNode] Done.";
```
Comment (floriantschopp) [inline applications/maplab-server-node/src/maplab-server-node.cc:232]:
```suggestion
    LOG(INFO) << "[MaplabServerNode] Done stopping SubmapProcessing threads.";
```

File: applications/maplab-server-node/src/maplab-server-node.cc [hunk 5]
```diff
@@ -207,29 +211,38 @@ void MaplabServerNode::start() {
...
 
-  LOG(INFO) << "[MaplabServerNode] Stopping Status thread...";
-  status_thread_.join();
-  LOG(INFO) << "[MaplabServerNode] Done.";
+  try {
+    LOG(INFO) << "[MaplabServerNode] Stopping Status thread...";
+    if (status_thread_.joinable()) {
+      status_thread_.join();
+    }
+    LOG(INFO) << "[MaplabServerNode] Done.";
```
Comment (floriantschopp) [inline applications/maplab-server-node/src/maplab-server-node.cc:242]:
```suggestion
    LOG(INFO) << "[MaplabServerNode] Done stopping Status thread.";
```

File: applications/maplab-server-node/test/test-maplab-server-node.cc [hunk 6]
```diff
@@ -35,10 +37,10 @@ class MaplabServerNodeTest : public ::testing::Test {
   const std::string kSubmap7 = kBasePath + "euroc_v1_01_submap_7";
 };
 
-TEST_F(MaplabServerNodeTest, TestMaplabServerNode) {
+TEST_F(MaplabServerNodeTest, DISABLED_TestMaplabServerNode) {
```
Comment (floriantschopp) [inline applications/maplab-server-node/test/test-maplab-server-node.cc:40]: I'm not happy with this 👎

File: applications/maplab-server-node/test/test-maplab-server-node.cc [hunk 7]
```diff
@@ -127,6 +129,7 @@ TEST_F(MaplabServerNodeTest, DISABLED_TestMaplabServerRosNodeLocal) {
   maplab_server_ros_node.shutdown();
 }
 
+
```
Comment (floriantschopp) [inline applications/maplab-server-node/test/test-maplab-server-node.cc:132]:
```suggestion
```

File: console-plugins/dense-mapping-constraints-plugin/include/dense-mapping/dense-mapping-parallel-process.h [hunk 8]
```diff
@@ -23,19 +23,24 @@ size_t parallelProcess(
     // Nothing todo.
     return 0u;
   }
+
+  const size_t num_processing_threads = std::min(num_elements, num_threads);
+  VLOG(1) << "num_processing_threads vs num_threads: " << num_processing_threads
```
Comment (floriantschopp) [inline console-plugins/dense-mapping-constraints-plugin/include/dense-mapping/dense-mapping-parallel-process.h:28]: Can we increase verbosity here?

File: applications/maplab-server-node/launch/arche/arche-maplab-server-test-base-station.launch [hunk 9]
```diff
@@ -41,7 +41,7 @@ Requires:
 		</include>
 	</group>
 
-	<group if="true">
+	<group if="false">
```
[the remark this answers is not in the document]
Comment (LBern) [reply applications/maplab-server-node/launch/arche/arche-maplab-server-test-base-station.launch:44]: I got so many `TF_old_data` messages such that I couldn't read anything on the console anymore. That's why I launched it separately.

File: applications/maplab-server-node/share/arche/arche-maplab-server-minimal.rviz (hunk 1 above)
Comment (LBern) [reply applications/maplab-server-node/share/arche/arche-maplab-server-minimal.rviz:208]: the glorious test artifacts ;)

File: console-plugins/dense-mapping-constraints-plugin/include/dense-mapping/dense-mapping-parallel-process.h (hunk 8 above)
Comment (LBern) [reply console-plugins/dense-mapping-constraints-plugin/include/dense-mapping/dense-mapping-parallel-process.h]: that was just debug actually -> I removed it

**Pull request merged**